File: Problem2971/solution.cpp
/* 

Time Complexity: O(nlog(n))
Space Complexity: O(1)

Algorithm with cpp std::sort() in O(n*log(n)) and greedy algorithm in O(n). In this version, the sum
is calculated before the greedy loop, so it is faster than the .py code.

*/

#include <vector>
#include <algorithm>
#include <iostream>

class Solution {
public:
    long long largestPerimeter(std::vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        unsigned long int s = 0;

        for(int j = 0; j < nums.size(); j++){
            s += nums[j];
        }
        for(int i = nums.size()-1; i >= 2; i--){
            s -= nums[i];
            if(s > nums[i]){
                return s+nums[i];
            }
        }
        return -1;
    }
};

int main(){

    Solution sol = Solution();
    std::vector<int> prob = {1,12,1,2,5,50,3};
    std::cout << sol.largestPerimeter(prob) << std::endl; // returns 12

}

File: Problem2971/solution.py
"""

Time Complexity: O(nlog(n))
Space Complexity: O(1)

Solution using the python sort, with n*log(n) complexity and greedy algorithm in O(n).
It was improved in the .cpp file

"""

class Solution:
    def largestPerimeter(self, nums: list[int]) -> int:
        s = 0
        nums.sort()
        
        for i in range(len(nums)-2):
            for j in range(len(nums)-i-1):
                s += nums[j]
            if s > nums[-i-1]:
                return s+nums[-i-1]
            s = 0
        return -1
    
sol:Solution = Solution()
prob:list[int] = [1,12,1,2,5,50,3]
print(sol.largestPerimeter(prob))